[5.0] New Feature - Change user before delete

degobbis opened the pull request:
### Summary of Changes
As a feature often requested in the German community and solved by me with a plugin, I was asked by @tecpromotion, @coolcat-creations and some others to bring the feature into the core. So here is my tribute.

It opens the possibility to define a user to be used as a replacement when another user is deleted. As long as this user is entered as a replacement, it cannot be deleted itself.

This affects extensions that use users, such as ```com_content```. Here it may also be necessary for copyright reasons that the original creator is still named. Therefore there is an option to enter the name of the user to be deleted as an alias.

All core tables in the database are searched for the user to be deleted and the user ID as well as the alias are adjusted according to the settings,

To make the whole thing also usable for other developers, a plugin group ```beforedeleteuser``` was added.
This makes it flexibly extendable, so that only an installable plugin with the necessary information is needed to implement the own extension.



### Testing Instructions
At this point I can only say, test all possible scenarios to delete a user. There are countless, so I will not list them all here.



### Actual result BEFORE applying this Pull Request
After deleting a user, the user ID remains in all core extensions, such as "com_content" and you get nasty warnings when editing an entry.



### Expected result AFTER applying this Pull Request
- No user can be deleted until a replacement user has been set.
- After configuration, the user ID of the deleted user is replaced by the ID of the replacement user.
- Depending on how it is set, the user name of the deleted user is entered as an alias. 
  Of course, only where this is also offered by the extension. 



### Known Issue:
For large websites it may timeout, but the user is not deleted then.
So far I have not been able to intercept this server message in such a way that a message is displayed which entries have already been processed and that the deletion has to be triggered again to process the rest.

Maybe someone has an idea how this can be solved.

degobbis committed:
Merge remote-tracking branch 'upstream/4.4-dev' into userdel/4.4-dev

degobbis committed:
Add the configuration tab to com_users

Add the configuration tab to com_users for the new functionality:
Change user before delete

degobbis committed:
New user plugin for 'Cahneg user before delete'

Add the user plugin for the new functionality to change the user ID on several extensions.
To allow 3rd party extensions to use this, a new plugin group is implementet.

degobbis committed:
Merge remote-tracking branch 'upstream/5.0-dev' into userdel/5.0-dev

degobbis committed:
Remove obsolete file

degobbis committed:
Check if the extension is enabled before loading it.

degobbis committed:
Add field validation

degobbis committed:
Add as core extension

degobbis committed:
Some error handlings

degobbis committed:
Merge remote-tracking branch 'upstream/5.0-dev' into userdel/5.0-dev

degobbis committed:
Merge remote-tracking branch 'upstream/5.0-dev' into userdel/5.0-dev

degobbis committed:
Merge remote-tracking branch 'joomla/5.0-dev' into userdel/5.0-dev

degobbis committed:
Merge remote-tracking branch 'joomla/5.0-dev' into userdel/5.0-dev

degobbis committed:
Typo

degobbis committed:
Fix - Alias was always entered

degobbis committed:
Added the verification if the fallback user exists any more

degobbis committed:
Removed some unneeded code and typo

degobbis committed:
Merge remote-tracking branch 'upstream/5.0-dev' into userdel/5.0-dev

degobbis committed:
Merge remote-tracking branch 'upstream/5.0-dev' into userdel/5.0-dev

degobbis commented:
I still have one question:
What must be done so that it is also taken into account when updating from Joomla 4 to Joomla 5?
Or rather who takes care of it?

brianteeman commented:
please review the submitted code to ensure that the copy pasted copyright and doc blocks have been updated for the submitted code

brianteeman commented:
please review the xml additions to ensure that they match the codestyle of the rest of the xml file

HLeithner commented:
> please review the submitted code to ensure that the copy pasted copyright and doc blocks have been updated for the submitted code

before you do this, we have to talk about the concept and the usecases and if they fit the needs of us and 3rd party extensions

brianteeman commented:
> > please review the submitted code to ensure that the copy pasted copyright and doc blocks have been updated for the submitted code
> 
> before you do this, we have to talk about the concept and the usecases and if they fit the needs of us and 3rd party extensions

in that case this is my comment to say that in its current form it should be an extension and not something for the core as in the real world this will be more of a pia than a problem solver. 

it doesnt take into account if the user has authored any content or even if the user was an approved user and not a fake spam registration

brianteeman commented:
it also doesn't work as it doesnt change the modified field

![image](https://github.com/joomla/joomla-cms/assets/1296369/20bf247e-966f-41fd-a177-5ce18ed43745)

coolcat-creations commented:
I would love to have such a feature because currently I struggle with lots of websites where I get the error that the user id was not found.
This can't be a core behaviour. When deleting an user the article should at least get the information that no user is assigned, but not an error "User Id Not found". It's possible now in core to assign "no user" so why is still the old user id assigned even the user is deleted?

sandewt commented:
In additioning to the previous comments:
Take a closer look at the _date_ , _version_ , _since_ data if they match what is common within Joomla. Use `__DEPLOY_VERSION__` if needed.

For some inspiration see #40553, there you will find good examples of the coding to be used.

Fedik commented:
Thanks for your work.

I have a 2 note about the feature.

**1. The reason**

The reason why people request such kind of feature is the message "user not found blabla", that is very annoying message.
The fix as easy as removing this line: https://github.com/joomla/joomla-cms/blob/d39717f0912051146a5d31aef3f19d350c3d1457/libraries/src/User/User.php#L823
This really should go, no idea why we keeping it.

**2. Not deleting user**

If we stilll want to prevent user from deleting. Then we should do what is com_privacy doing: rename user data to random values. Instead of deleting and then creating new user with a loot of overhead for replacing old IDs.

degobbis commented:
I am not a professional in programming, so this is not the "Eierlegende Wollmilchsau" (as we say in Germany). 
However, if com_privacy provides that, maybe we should put the developer on it.

This feature is just to implement the function that Worpress already has for a long time.

When a user is deleted and he is used in pages or posts, a replacement user must be selected.

For Joomla just a little more elaborate, but in principle the same concept.
I could not find a process for this in Joomla, hence my PR.

However, if the feature is not needed, or does not meet the wishes of all, I hope it finds a professional who can implement it.

I have reached the limits of my abilities.
Gladly someone from the German-speaking area can get in touch, who knows all Joomla interfaces and who can take the time to make this better with me.

I would be happy, because I like to learn.

coolcat-creations commented:
I would be very happy to have a solution for this and I am very glad you like to help in Joomla Core because you have great solutions all over the place. I think maybe a good thing would be to have a "delete user" which does not delete the user but all his personal data, and a permanent delete user which is only possible when all his data is reassigned?

sandewt commented:
I have an additional wish that if a user has been inactive for a certain period of time (for example a year), this user will automatically be anonymized and his data will be deleted.

coolcat-creations commented:
> I have an additional wish that if a user has been inactive for a certain period of time (for example a year), this user will automatically be anonymized and his data will be deleted.

Sounds great and could be maybe managed with com privacy?

sandewt commented:
>  ....maybe managed with com privacy?

Indeed, GDPR reasons.

degobbis commented:
> I have an additional wish that if a user has been inactive for a certain period of time (for example a year), this user will automatically be anonymized and his data will be deleted.

In my opinion this would be more like a scheduler plugin

degobbis committed:
Merge remote-tracking branch 'upstream/5.0-dev' into userdel/5.0-dev

richard67 commented:
In case if it helps I have prepared a list of all table columns in 5.0-dev which contain a user ID:

```
TABLE_NAME               COLUMN_NAME
#__action_logs           user_id
#__action_logs_users     user_id
#__banner_clients        checked_out
#__banners               checked_out
#__banners               created_by
#__banners               modified_by
#__categories            checked_out
#__categories            created_user_id
#__categories            modified_user_id
#__contact_details       checked_out
#__contact_details       created_by
#__contact_details       modified_by
#__contact_details       user_id
#__content               checked_out
#__content               created_by
#__content               modified_by
#__extensions            checked_out
#__fields                checked_out
#__fields                created_user_id
#__fields                modified_by
#__fields_groups         checked_out
#__fields_groups         created_by
#__fields_groups         modified_by
#__finder_filters        checked_out
#__finder_filters        created_by
#__finder_filters        modified_by
#__guidedtour_steps      checked_out
#__guidedtour_steps      created_by
#__guidedtour_steps      modified_by
#__guidedtours           checked_out
#__guidedtours           created_by
#__guidedtours           modified_by
#__history               editor_user_id
#__menu                  checked_out
#__messages              user_id_from
#__messages              user_id_to
#__messages_cfg          user_id
#__modules               checked_out
#__newsfeeds             checked_out
#__newsfeeds             created_by
#__newsfeeds             modified_by
#__privacy_consents      user_id
#__scheduler_tasks       checked_out
#__scheduler_tasks       created_by
#__session               userid
#__tags                  checked_out
#__tags                  created_user_id
#__tags                  modified_user_id
#__ucm_content           core_checked_out_user_id
#__ucm_content           core_created_user_id
#__ucm_content           core_modified_user_id
#__update_sites          checked_out
#__user_mfa              user_id
#__user_notes            checked_out
#__user_notes            created_user_id
#__user_notes            modified_user_id
#__user_notes            user_id
#__user_profiles         user_id
#__user_usergroup_map    user_id
#__workflow_stages       checked_out
#__workflow_transitions  checked_out
#__workflows             checked_out
#__workflows             created_by
#__workflows             modified_by
```
The "checked_out" might be not relevant since there should nothing be checked out by a not existing user, and if so, a global checkin should handle that.

degobbis committed:
Completely reworked

The user plugin has been removed and introduced as a trait for the plugins in the "beforedeleteuser" plugin group. This gives the developers of third party extensions more flexibility to intervene in the exchange process themselves.

All core extensions are now combined into one "beforedelteuser" plugin.
"Modified By" is now also taken into consideration.

The global configuration of the settings has been removed, instead a popup is opened before deletion and the required settings are requested.

All adjustments could only be realized because bembelimen gave me some good tips.

Surely there is still room for improvement, but I hope you are a bit proud of me ;-)

degobbis committed:
Merge remote-tracking branch 'upstream/5.0-dev' into userdel/5.0-dev

degobbis commented:
Completely reworked.

The user plugin has been removed and introduced as a trait for the plugins in the "beforedeleteuser" plugin group. This gives the developers of third party extensions more flexibility to intervene in the exchange process themselves.

All core extensions are now combined into one "beforedelteuser" plugin.
"Modified By" is now also taken into consideration.

The global configuration of the settings has been removed, instead a popup is opened before deletion and the required settings are requested.

All adjustments could only be realized because @bembelimen gave me some good tips.

Surely there is still room for improvement, but I hope you are a bit proud of me ;-)

brianteeman commented on administrator/components/com_users/forms/before_delete.xml:
```diff
@@ -0,0 +1,39 @@
+<form>
+	<fieldset name="before_delete_user"
+	          label="COM_USERS_BEFORE_DELETE_USER">
+		<field name="fallbackUserIdOnDelete"
```
please sort the values in the following order
name
type
label
description
and then everything else

brianteeman commented on administrator/language/en-GB/com_users.ini:
```diff
@@ -19,6 +19,21 @@ COM_USERS_BATCH_DELETE="Delete From Group"
 COM_USERS_BATCH_GROUP="Select Group"
 COM_USERS_BATCH_OPTIONS="Batch process the selected users"
 COM_USERS_BATCH_SET="Move To Group"
+COM_USERS_BEFORE_DELETE_USER="Before Delete User"
```
Please make sure that language strings are alpha sorted

brianteeman commented on administrator/language/en-GB/plg_beforedeleteuser_core.ini:
```diff
@@ -0,0 +1,7 @@
+; Joomla! Project
+; (C) 2016 Open Source Matters, Inc. <https://www.joomla.org>
```
please use the correct year for all copyright dates. If it is a new fil then it should be 2023

brianteeman commented on installation/sql/postgresql/base.sql:
```diff
@@ -264,6 +264,19 @@ INSERT INTO "#__extensions" ("package_id", "name", "type", "element", "folder",
 (0, 'plg_authentication_joomla', 'plugin', 'joomla', 'authentication', 0, 1, 1, 1, 1, '', '', '', 2, 0),
 (0, 'plg_authentication_ldap', 'plugin', 'ldap', 'authentication', 0, 0, 1, 0, 1, '', '{"host":"","port":"389","use_ldapV3":"0","negotiate_tls":"0","no_referrals":"0","auth_method":"bind","base_dn":"","search_string":"","users_dn":"","username":"admin","password":"bobby7","ldap_fullname":"fullName","ldap_email":"mail","ldap_uid":"uid"}', '', 3, 0),
+(0, 'plg_beforedeleteuser_core', 'plugin', 'core', 'beforedeleteuser', 0, 1, 1, 1, 1, '', '', '', 1, 0),
+(0, 'plg_beforedeleteuser_categories', 'plugin', 'categories', 'beforedeleteuser', 0, 1, 1, 1, 1, '', '', '', 1, 0),
```
did you forget to remove these?

degobbis committed:
Manifest values sorted

degobbis committed:
Alpha sorted language strings

degobbis committed:
Forgot to remove the old plugin entries from the SQL

degobbis committed:
Updated every copyright year and unified the headers for the new files

brianteeman commented on administrator/components/com_users/tmpl/users/default_before_delete_body.php:
```diff
@@ -0,0 +1,27 @@
+ * @package     Joomla.Administrator
+ * @subpackage  com_users
+ *
+ * @copyright   (C) 2015 Open Source Matters, Inc. <https://www.joomla.org>
```
2023

degobbis replied:
Fixed, thx

brianteeman commented on administrator/components/com_users/tmpl/users/default_before_delete_footer.php:
```diff
@@ -0,0 +1,21 @@
+ * @package     Joomla.Administrator
+ * @subpackage  com_users
+ *
+ * @copyright   (C) 2015 Open Source Matters, Inc. <https://www.joomla.org>
```
2023

degobbis replied:
Fixed

brianteeman commented on administrator/language/en-GB/com_users.ini:
```diff
@@ -19,6 +19,21 @@ COM_USERS_BATCH_DELETE="Delete From Group"
+COM_USERS_BEFORE_DELETE_USER_FIELD_FALLBACK_USER_OPTION_EMPTY="Select user"
+COM_USERS_BEFORE_DELETE_USER_FIELD_SET_ALIAS_DESC="Write the name of the user to be deleted in the 'Author alias' field if it is empty."
+COM_USERS_BEFORE_DELETE_USER_FIELD_SET_ALIAS_LABEL="Enter author alias"
+COM_USERS_BEFORE_DELETE_USER_FIELD_OVERRIDE_ALIAS_DESC="Overwrite the content in the 'Author alias' field with the name of the user to be deleted, if this is not empty."
```
please alpha sort

degobbis replied:
Fixed

brianteeman commented on libraries/src/Extension/ExtensionHelper.php:
```diff
@@ -171,6 +171,18 @@ class ExtensionHelper
         ['plugin', 'ldap', 'authentication', 0],
 
+        // Core plugin extensions - beforedeleteuser
+        ['plugin', 'banners', 'beforedeleteuser', 0],
```
did you forget to update this?

degobbis replied:
Fixed

brianteeman commented on administrator/components/com_users/src/Controller/UsersController.php:
```diff
@@ -170,4 +176,98 @@ public function getQuickiconContent()
+     *
+     * @return  void
+     *
+     * @since   1.6
```
* @since   __DEPLOY_VERSION__

degobbis replied:
This is not from me...

degobbis replied:
Ahhh.... the delete :-)
Fixed

degobbis committed:
Thx Brianteeman

brianteeman commented on administrator/components/com_users/tmpl/users/default_before_delete_body.php:
```diff
@@ -0,0 +1,26 @@
+<?php
+
+/**
+ * Joomla! Content Management System
```
replace this line with

```
 * @package     Joomla.Administrator
 * @subpackage  com_users
```

brianteeman commented on administrator/components/com_users/tmpl/users/default_before_delete_footer.php:
```diff
@@ -0,0 +1,20 @@
+<?php
+
+/**
+ * Joomla! Content Management System
```
replace this line with

```
 * @package     Joomla.Administrator
 * @subpackage  com_users
```

brianteeman commented on administrator/components/com_users/tmpl/users/default_before_delete_footer.php:
```diff
@@ -0,0 +1,20 @@
+use Joomla\CMS\Language\Text;
+
+?>
+<button type="button" class="btn btn-secondary" aria-label="<?php echo Text::_('JCANCEL'); ?>" data-bs-dismiss="modal">
```
no need for the aria-label here as the text is displayed.

brianteeman commented on administrator/language/en-GB/com_users.ini:
```diff
@@ -19,6 +19,21 @@ COM_USERS_BATCH_DELETE="Delete From Group"
 COM_USERS_BATCH_OPTIONS="Batch process the selected users"
 COM_USERS_BATCH_SET="Move To Group"
+COM_USERS_BEFORE_DELETE_USER="Before Delete User"
+COM_USERS_BEFORE_DELETE_USER_CHANGED_FALLBACK_ALIAS_MSG="The autoralias <code>%s</code> was deposited."
```
autoralias ==> author alias

not sure what you mean by deposited

degobbis replied:
Translated with deepl:
wurde hinterlegt == was deposited
wurde deponiert == was deposited
wurde eingesetzt == was placed
wurde eingetragen == was inserted
What du You think is better?

brianteeman replied:
For now please use inserted. It can always be refined later

brianteeman commented on administrator/language/en-GB/com_users.ini:
```diff
@@ -19,6 +19,21 @@ COM_USERS_BATCH_DELETE="Delete From Group"
+COM_USERS_BEFORE_DELETE_USER_ERROR_FALLBACK_USER_NOT_SET_MSG="The fallback user has not been set. Please select a fallback user to process the deleting."
+COM_USERS_BEFORE_DELETE_USER_ERROR_USER_NOT_DELETED_MSG="No user was deleted!"
+COM_USERS_BEFORE_DELETE_USER_FIELD_FALLBACK_USER_DESC="Select user to be put in place of the deleted user."
+COM_USERS_BEFORE_DELETE_USER_FIELD_FALLBACK_USER_LABEL="Fallback user"
```
Please update all labels so be capitalised eg Fallback User

brianteeman commented on libraries/src/Form/Form.php:
```diff
@@ -617,6 +617,8 @@ public function load($data, $replace = true, $xpath = null)
         // Get the XML elements to load.
         $elements = [];
 
+        $test = $data->getName();
```
what is this for?

degobbis replied:
Ups, placed for debugging :-)
Removed

brianteeman commented on plugins/beforedeleteuser/core/services/provider.php:
```diff
@@ -0,0 +1,50 @@
+<?php
+
+/**
+ * Joomla! Content Management System
```
replace this line with

```
 * @package     Joomla.Plugin
 * @subpackage  beforedeleteuser.core
```

brianteeman commented on plugins/beforedeleteuser/core/src/Extension/Core.php:
```diff
@@ -0,0 +1,272 @@
+<?php
+
+/**
+ * Joomla! Content Management System
```
replace this line with

```
 * @package     Joomla.Plugin
 * @subpackage  beforedeleteuser.core
```

brianteeman commented on plugins/beforedeleteuser/core/src/Extension/Core.php:
```diff
@@ -0,0 +1,272 @@
+use Joomla\Database\DatabaseAwareTrait;
+use Joomla\Event\Event;
+use Joomla\Event\SubscriberInterface;
+use Joomla\CMS\User\BeforeDeleteUserInterface;
```
please alpha sort

brianteeman commented on plugins/beforedeleteuser/core/src/Extension/Core.php:
```diff
@@ -0,0 +1,272 @@
+use Joomla\CMS\User\BeforeDeleteUserInterface;
+
+/**
+ * Class to support the core extension 'com_content'.
```
com_content?

brianteeman commented on plugins/beforedeleteuser/core/services/provider.php:
```diff
@@ -0,0 +1,50 @@
+     *
+     * @return  void
+     *
+     * @since   4.2.0
```
`* @since __DEPLOY_VERSION__`

brianteeman commented:
Please review the errors reported in the php cs by drone https://ci.joomla.org/joomla/joomla-cms/67021/1/7

degobbis committed:
Corrections

degobbis commented:
@brianteeman Thx for for your effort!

brianteeman commented on administrator/language/en-GB/com_users.ini:
```diff
@@ -19,6 +19,21 @@ COM_USERS_BATCH_DELETE="Delete From Group"
 COM_USERS_BATCH_OPTIONS="Batch process the selected users"
 COM_USERS_BATCH_SET="Move To Group"
+COM_USERS_BEFORE_DELETE_USER="Before Delete User"
+COM_USERS_BEFORE_DELETE_USER_CHANGED_FALLBACK_ALIAS_MSG="The autor alias <code>%s</code> was deposited."
```
author

brianteeman commented on administrator/language/en-GB/com_users.ini:
```diff
@@ -19,6 +19,21 @@ COM_USERS_BATCH_DELETE="Delete From Group"
 COM_USERS_BATCH_SET="Move To Group"
+COM_USERS_BEFORE_DELETE_USER="Before Delete User"
+COM_USERS_BEFORE_DELETE_USER_CHANGED_FALLBACK_ALIAS_MSG="The autor alias <code>%s</code> was deposited."
+COM_USERS_BEFORE_DELETE_USER_CHANGED_FALLBACK_ALIAS_IF_NOT_EMPTY_MSG="The autor alias <code>%s</code> was deposited, if the field was empty before."
```
author

brianteeman commented on plugins/beforedeleteuser/core/core.xml:
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<extension type="plugin" group="behaviour" method="upgrade">
```
is this the correct group?

degobbis committed:
Corrections and drone fixes

degobbis committed:
typo deposited => inserted

degobbis committed:
Fix phpcs

Fedik commented:
I made alternative fix, please test #41048

degobbis committed:
Merge remote-tracking branch 'joomla/5.0-dev' into userdel/5.0-dev

degobbis committed:
Merge remote-tracking branch 'joomla/5.0-dev' into userdel/5.0-dev

degobbis committed:
Merge branch '5.0-dev' into userdel/5.0-dev

degobbis committed:
Merge branch '5.0-dev' into userdel/5.0-dev

degobbis committed:
Merge remote-tracking branch 'upstream/5.0-dev' into userdel/5.0-dev

HLeithner commented:
This pull request has been automatically rebased to 5.1-dev.

jjpcoca commented:
I have tested it but it does not work correctly. The deleted user is still assigned to the article with the ID but does not replace it with the chosen one.

![screen shot 2023-10-05 at 12 50 47](https://issues.joomla.org/uploads/1/f6a56bc2e543c4d270be057ab2586406.jpg)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40751">issues.joomla.org/tracker/joomla-cms/40751</a>.</sub>

degobbis committed:
Merge remote-tracking branch 'upstream/5.0-dev' into userdel/5.0-dev

degobbis committed:
Merge remote-tracking branch 'upstream/5.1-dev' into userdel/5.0-dev

Hackwar commented:
Can you finish this, so that we can test this during PBF?

coolcat-creations commented:
@degobbis I would love the Feature in 5.1 :-)

degobbis committed:
Merge remote-tracking branch 'upstream/5.1-dev' into userdel/5.0-dev

rdeutz commented:
I am closing this, we discussed this today and came to the conclusion that we want to go a different way. We think it is better to anonymise the user data instead of hard delete a user account. With this we not only fix it for com_content, also other extensions don't run into the problem of not existing user account. Essentially we do what we do in other areas and trash user accounts instead of delete. There will be an option for remove a user from the database, this will come with a warning that there can be site effects.

Thank you Guido for proposing this feature and working on this.

degobbis commented:
Too bad, I just finished the fix, including the action log.

ot2sen commented:
@rdeutz Will this different way handle the Author rights (copyrights as said in PR intro)?
This PR using an Alias to preserve these rights did seem to provide a solution to that, although it didn´t seem necessary to go as far as always have a blanco user (spam regs and such.)
